common/src/patch.rs

```diff
@@ -0,0 +1,219 @@
+//! Patch-related functions and types.
+use anyhow::{Error, Result};
+use git2::Oid;
+use regex::Regex;
+
+use librad::git::refs::Refs;
+use librad::git::storage::{ReadOnly, ReadOnlyStorage};
+
+use crate::{git, person, project};
+use rad_terminal::components as term;
+
+pub const TAG_PREFIX: &str = "patches/";
+
+#[derive(PartialEq, Eq)]
+pub enum State {
+    Open,
+    Merged,
+}
+
+/// A patch is a change set that a user wants the maintainer to merge into a project's default
+/// branch.
+///
+/// A patch is represented by an annotated tag, prefixed with `radicle-patch/`.
+pub struct Metadata {
+    /// ID of a patch. This is the portion of the tag name following the `radicle-patch/` prefix.
+    pub id: String,
+    /// Peer that the patch originated from
+    pub peer: project::PeerInfo,
+    /// Message attached to the patch. This is the message of the annotated tag.
+    pub message: Option<String>,
+    /// Head commit that the author wants to merge with this patch.
+    pub commit: Oid,
+}
+
+impl Metadata {
+    pub fn name(&self) -> String {
+        str::replace(&self.id, TAG_PREFIX, "")
+    }
+}
+
+/// List patches on the local device. Returns a given peer's patches or this peer's
+/// patches if `peer` is `None`.
+pub fn list<S>(
+    storage: &S,
+    peer: Option<project::PeerInfo>,
+    project: &project::Metadata,
+) -> Result<Vec<Metadata>, Error>
+where
+    S: AsRef<ReadOnly>,
+{
+    let storage = storage.as_ref();
+    let mut patches: Vec<Metadata> = vec![];
+
+    let peer_id = peer.clone().map(|p| p.id);
+    let info = match peer {
+        Some(info) => info,
+        None => self_info(storage, project)?,
+    };
+
+    if let Ok(refs) = Refs::load(&storage, &project.urn, peer_id) {
+        let blobs = match refs {
+            Some(refs) => refs.tags().collect(),
+            None => vec![],
+        };
+        for (_, oid) in blobs {
+            match storage.find_object(oid) {
+                Ok(Some(object)) => {
+                    let tag = object.peel_to_tag()?;
```

Comment:
Calling `object.into_tag` results in the compiler error `libgit2_sys::git_repository cannot be shared between threads safely`. Shall we look into this seperately, since I could not resolve that easily?

Reply:
Yup.